toctoc/slickMySql/src/main/scala/io.buildo.toctoc/slick/authentication/login/MySqlSlickLoginAuthenticationDomain.scala

```diff
@@ -38,41 +31,36 @@ class MySqlSlickLoginAuthenticationDomain[F[_]: FutureLift[*[_], Future]](
   }
   val loginTable = TableQuery[LoginTable]
 
-  override def register(s: Subject, c: Login): F[Either[AuthenticationError, LoginDomain[F]]] =
-    F.delay {
+  override def register(s: Subject, c: Login): IO[AuthenticationError, LoginDomain] =
+    ZIO.fromFuture { _ =>
       db.run(loginTable += ((0, s.ref, c.username, hashPassword(c.password))))
-    }.futureLift
-      .as(this.asRight[AuthenticationError])
-      .handleError { case _ =>
-        AuthenticationError.InvalidCredential.asLeft
-      }
-      .widen
+    }.map(_ => this).mapError(_ => AuthenticationError.InvalidCredential)
 
-  override def unregister(s: Subject): F[Either[AuthenticationError, LoginDomain[F]]] =
-    F.delay {
+  override def unregister(s: Subject): IO[AuthenticationError, LoginDomain] =
+    ZIO.fromFuture { _ =>
       db.run(loginTable.filter(_.ref === s.ref).delete)
-    }.futureLift.as(this.asRight)
+    }.map(_ => this).mapError(_ => AuthenticationError.Forbidden)
 
-  override def unregister(c: Login): F[Either[AuthenticationError, LoginDomain[F]]] =
-    (for {
-      a <- EitherT(authenticate(c))
+  override def unregister(c: Login): IO[AuthenticationError, LoginDomain] =
+    for {
+      a <- authenticate(c)
       (_, s) = a
-      res <- EitherT(unregister(s))
-    } yield res).value
+      res <- unregister(s)
+    } yield res
 
-  override def authenticate(c: Login): F[Either[AuthenticationError, (LoginDomain[F], Subject)]] = {
-    F.delay {
+  override def authenticate(c: Login): IO[AuthenticationError, (LoginDomain, Subject)] = {
```

Comment:
Ho qualche dubbio su come si comporti questa funzione (e le altre di questo file e altri con DB) in caso di errori imprevisti, tipo timeout di connessione.

A occhio la versione in ZIO dà `AuthenticationError.InvalidCredential` perché `ZIO.fromFuture` mappa ogni `Future` fallita in un fail (non un die) e poi c'è il `mapError`.

Nel caso del codice TF invece mi chiedo se finisca in un fallimento della `F` e non in un `AuthenticationError`. In quel caso l'equivalente ZIO magari sarebbe fare un `.orDie` per errori generici e mappare ad `AuthenticationError` solo i casi in cui non si trovano le credenziali.

Non sono sicuro della differenza però perché magari mi perdo qualcosa di `delay`/`futureLift`/ecc. quindi chiederei solo di controllare (per questa e le altre funzioni con DB) e vedere se ci sono differenze e in caso se ci tornano.